lib/src/git_subprocess.rs

```diff
@@ -0,0 +1,589 @@
+fn parse_no_remote_ref(stderr: &[u8]) -> Option<String> {
+    let first_line = stderr.lines().next()?;
+    first_line
+        .strip_prefix(b"fatal: couldn't find remote ref ")
+        .map(|refname| refname.to_str_lossy().into_owned())
+}
+
+/// Parse remote tracking branch not found
+///
+/// This returns true if the error was detected
+///
+/// if a branch is asked for but is not present, jj will detect it post-hoc
+/// so, we want to ignore these particular errors with git
+///
+/// The first line is of the form:
+/// `error: remote-tracking branch '<branch>' not found`
+fn parse_no_remote_tracking_branch(stderr: &[u8]) -> Option<String> {
+    let first_line = stderr.lines().next()?;
+
+    let suffix = first_line.strip_prefix(b"error: remote-tracking branch '")?;
+
+    suffix
+        .strip_suffix(b"' not found.")
+        .or(suffix.strip_suffix(b"' not found"))
+        .map(|branch| branch.to_str_lossy().into_owned())
+}
+
+// return the fully qualified ref that failed to fetch
+//
+// note that git fetch only returns one error at a time
+fn parse_git_fetch_output(output: Output) -> Result<Option<String>, GitSubprocessError> {
+    if output.status.success() {
+        return Ok(None);
+    }
+
+    // There are some git errors we want to parse out
+    if let Some(remote) = parse_no_such_remote(&output.stderr) {
+        return Err(GitSubprocessError::NoSuchRepository(remote));
+    }
+
+    if let Some(refspec) = parse_no_remote_ref(&output.stderr) {
+        return Ok(Some(refspec));
+    }
+
+    if parse_no_remote_tracking_branch(&output.stderr).is_some() {
+        return Ok(None);
+    }
+
+    Err(external_git_error(&output.stderr))
+}
+
+fn parse_git_branch_prune_output(output: Output) -> Result<(), GitSubprocessError> {
+    if output.status.success() {
+        return Ok(());
+    }
+
+    // There are some git errors we want to parse out
+    if parse_no_remote_tracking_branch(&output.stderr).is_some() {
+        return Ok(());
+    }
+
+    Err(external_git_error(&output.stderr))
+}
+
+fn parse_git_remote_show_output(output: Output) -> Result<Output, GitSubprocessError> {
+    if output.status.success() {
+        return Ok(output);
+    }
+
+    // There are some git errors we want to parse out
+    if let Some(remote) = parse_no_such_remote(&output.stderr) {
+        return Err(GitSubprocessError::NoSuchRepository(remote));
+    }
+
+    Err(external_git_error(&output.stderr))
+}
+
+fn parse_git_remote_show_default_branch(
+    stdout: &[u8],
+) -> Result<Option<String>, GitSubprocessError> {
+    Ok(stdout
+        .to_str()
+        .map_err(|e| {
+            GitSubprocessError::External(format!("git remote output is not utf-8: {e:?}"))
+        })?
+        .lines()
+        .map(|x| x.trim())
+        .find(|x| x.starts_with("HEAD branch:"))
+        .and_then(|x| x.split(" ").last().map(|y| y.trim().to_string()))
+        .filter(|branch_name| branch_name != "(unknown)"))
```

Comment:
nit: maybe we can rewrite this to parse `&[u8]` and convert to `String` at the end?